lib/src/iterable_extensions.dart

```diff
@@ -16,6 +18,37 @@ import 'algorithms.dart';
 /// iterables with specific element types include those of
 /// [IterableComparableExtension] and [IterableNullableExtension].
 extension IterableExtension<T> on Iterable<T> {
+  /// Chooses [count] elements at random from this iterable.
+  ///
+  /// There must be at least [count] elements in the iterable.
+  /// The returned list contains [count] different elements
+  /// of the iterable.
+  /// If the same value occurs more than once in the iterable,
+  /// it can also occur more than once in the chosen elements.
+  ///
+  /// Each element of the iterable has the same chance of
+  /// being chosen.
+  /// The chosen elements are not in any specific order.
+  List<T> choose(int count, [Random? random]) {
```

Comment:
This variant throws if there are too few.

Should we also have one which just returns all the elements if there are fewer than `count`?
Say `chooseMax(int count, [Random? random])`.

Or make that the default behavior?

(I don't want a parameter to choose between the two behaviors in one function, that's unnecessarily messy, and I dislike using booleans to switch between different algorithms).

Reply:
I also dislike the idea of a boolean argument.

I'm not excited about needing two methods for this, but I don't have a sense about which is more useful. The version that _doesn't_ throw can be used as a basis for the version that does throw, so it's more general, but also adds noise in the code if that isn't the behavior that the user was looking for.

I checked other languages/libraries and I couldn't find this method implemented in many of them. The python version throws, while the ruby version returns as many elements as it can (interestingly it also shuffles them).

While I was looking though, it did occur to me that most readers might find "sample" more intuitive than "choose". Even I associate "choose" more with the combinatorics operation that only considers the size, not the specific elements.

WDYT about renaming to "sample", and going with the version that does not throw for now. If we find usage is high enough, and enough folks are writing `if (sample.length < desiredLength) throw` we could add a second version.